### ISSUE TITLE
Random sampling of k elements from a dask bag

### ISSUE THREAD

**Issue by mathdugre** [dask/dask#4799]
It seems like there is no way, from the API, to randomly sample k elements from a dask bag.

#1332 Address a similar issue. It gives a random probability to each elements to be chosen however you are not guaranteed to have a fix number of element at the end.

Currently the way I would do it is:

```
import dask.bag as db
import random

data = db.from_sequence([1, 2, 3, 4, 5, 6])
random.sample(data, 3)  # [2, 4, 5]
```

What I think a `sample` method from `dask.bag` should do:

```
import dask.bag as db

data = db.from_sequence([1, 2, 3, 4, 5, 6])
data.sample(k)  # [2, 3, 6]
```

If this look like a desirable feature for the API I could try working on it during the next few weeks.

**Comment by jrbourbeau** [comment]
Thanks @mathdugre for opening up this issue and for the clear code snippets. @mrocklin do you have any thoughts on adding a `sample` method to the dask bag API?

**Comment by mrocklin** [comment]
In principle I think it would be good to have and the proposed API seems fine to me, especially given that it mirrors the standard library.

I think that it might be somewhat challenging to implement because we don't know how many elements are in each partition.  My guess is that it's not trivial to fairly select k elements from multiple partitions lazily without knowing how many elements are in each partition.  I also suspect that it's possible, but may require some cleverness.

**Comment by martindurant** [comment]
A function to downsample bags by randomly taking 1-in-N elements would be much easier.

**Comment by jrbourbeau** [comment]
To summarize, it sounds like a `sample` method would be a welcome addition to the bag API. However, it's possible that it may not be straightforward to implement. We're happy to help review and provide feedback @mathdugre

**Comment by mathdugre** [comment]
I will slowly try to work on it. It will probably take a bit of time initially since I will need to familiarize myself with the code base. I will let you know if I have any issues; thank you.

**Comment by jrbourbeau** [comment]
Thank you!

**Comment by martindurant** [comment]
We do already have `dask.bag.Bag.random_sample`, is that enough?

**Comment by jrbourbeau** [comment]
I think @mathdugre's concern is that with `random_sample` you're not guaranteed to have a fix number of elements

**Comment by mathdugre** [comment]
My concern is the one mentioned by @jrbourbeau. In some case you would want to sample an exact number of element and I think it would be hard to implement this through `dask.bag.Bag.ramdom_sample`

**Comment by martindurant** [comment]
Understood, but I don't think exact number with fair sampling can be done in parallel, because you don't know how many elements there are in each partition.

**Comment by mrocklin** [comment]
One approach would be to gather k elements from each partition as well as
the total number of elements from each partition, and then sample k
elements from that group in some weighted way.  This probably gets awkward
if partitions have fewer than k elements or if k is very large.

> but I don't think exact number with fair sampling can be done in parallel

Sampling algorithms can be very clever sometimes.  My solution above is
probably overly naive in some way.  I expect that many things can be done,
they just won't be trivial.  As an example, have you seen the numerically
precise single pass moment (std/var) calculation that @jcrist implemented
for dask arrays?  There are a lot of fun parallel algorithms out there if
people want to pursue them.

**Comment by mrocklin** [comment]
Checking in here.  @mathdugre is this something that you're still interested in doing?  If so, is there anything that you need in order to make progress?

**Comment by mathdugre** [comment]
@mrocklin Yes I'm still interested in this issue. I didn't have much time in the last two weeks but I should be able to take a closer look during next weekend.

**Comment by eracle** [comment]
Hey guys, really nice project here.
Imho sampling requires each element to have equal probability to being selected.
As pandas does in [this line of code](https://github.com/pandas-dev/pandas/blob/v1.0.3/pandas/core/generic.py#L5034).
So, knowing the total number of elements in the bag is a mandatory requirement.

It seems strange to me that the bag data structure doesn't have a lazy attribute count (or len) that is updated each time any sort of change is done on the bag. 

By having it, implementing the sample functinality would result to be straightforward, as is implemented on pandas.
Frankly speaking I would love to contribute on this.

**Comment by TomAugspurger** [comment]
> It seems strange to me that the bag data structure doesn't have a lazy attribute count (or len) that is updated each time any sort of change is done on the bag.

We wouldn't be able to know the partition lengths when the result depends on the value of the data (e.g. `bag.filter(lambda x: x > 0)`) since we would have to run the graph to determine the lengths.

>  Frankly speaking I would love to contribute on this.

I think we're happy to have a solution here. Earlier there was a suggestion to gather `k` elements and the lengths of each partition and second sample weighted by the number of elements in each partition. It wouldn't be perfectly uniform sampling, but I think would be acceptable for most users.

**Comment by martindurant** [comment]
You can also sample uniformly if you process the data serially, having the probability of inclusion depend on the number of elements seen so far. For the final node of a graph, accumulating a final result, that might be useful.

**Comment by mrocklin** [comment]
> Earlier there was a suggestion to gather k elements and the lengths of
each partition and second sample weighted by the number of elements in each
partition. It wouldn't be perfectly uniform sampling, but I think would be
acceptable for most users.

FWIW, this seems like the right approach to me.  I don't immediately see
how it wouldn't be perfectly uniform.

**Comment by eracle** [comment]
Ok, I take this issue. I am already on the way.
Soon I will publish a PR and also an explanation of the proposed solution.

**Comment by eracle** [comment]
Hello there, I made the PR.
You can find a brief description of the proposed solution [here](https://medium.com/@eracle/distributed-sampling-on-dask-bags-f0e9faad8610).

**Comment by eracle** [comment]
Ok, I think we can successfully close this.

**Comment by jcrist** [comment]
Fixed by #6208.
